### ISSUE TITLE
Forwarding connect message to another channel and session creation

### ISSUE THREAD

**Issue by raphael-boucher** [django/channels#356]
I'm forwarding the connection message from a channel that creates the session to another channel that suppose the session exists, like this:

``` python
class AnotherConsumer(BaseConsumer):
    # get the session, that needs to be already created
    channel_user_session = True

    method_mapping = {
        'another.connect': 'raw_connect',
    }

class MyConsumer(JSONWebsocketConsumer):
    # create the session from http cookies sent on connect message
    http_user = True

    def connect(message):
         Channel('another').send(message.content)
```

When using `runserver`, the session is lost in the second consumer most of the time (it works randomly and rarely).
When using `runserver --no_worker` + `runworker` or setting a breaking point just before the fowarding, it works fine all the time.

A solution I have right now is to create the session again on the second consumer, but I would like to avoid that if possible.
I suppose the session is not fully registered when the message is forwarded to the second channel in `runserver` mode. Any idea how to ensure that ?

**Comment by andrewgodwin** [comment]
Can you say what you mean by "the session is lost"? I suspect you're encountering an ordering problem (see http://channels.readthedocs.io/en/latest/getting-started.html#enforcing-ordering) and the other consumer is being called before your one exits, so the session is not yet persisted.

I'm not sure quite how you would fix this, given that you cannot use `enforce_ordering`. Maybe we need a decorator that delays sends till function completion (and the session is written), or maybe we should write the session before connect is called.

**Comment by raphael-boucher** [comment]
It feels like the session is lost since it exists in the first consumer, but it was a poor formulation.
The session is actually not yet created, for the reason you just said:

```
#in channel_session

  # Run the consumer
        try:
            return func(message, *args, **kwargs)
        finally:
            # Persist session if needed
            if session.modified:
                session.save()
```

Is there a reason to save the session only after connect is called?

**Comment by andrewgodwin** [comment]
Well, it needs to be saved at the end in case it's been changed, but we could save it before as well.

**Comment by raphael-boucher** [comment]
Fair enough!

**Comment by eternaldensity** [comment]
I encountered this problem too.
Basically, while the transfer_user function executes before the connect consumer, the session.save doesn't execute until after the connect consumer.  So it's not guaranteed to have saved the transfer before any other consumers (that are sent messages from the connect consumer) get executed (or more specifically, before channel_session invokes for them.)
I noticed this because on occasion I would get the user come up as AnonymousUser in consumers invoked from the connect consumer, even though the page's view required login, and the connect consumer had already confirmed that the user was authenticated before invoking the other consumers.  I confirmed that this was definitely the issue by putting a time.sleep(2) at the end of my connect consumer.  That forced the issue to always happen.  (It's a good way to test concurrency issues.)

But sticking message.channel_session.save() before invoking the other consumers fixes it quite definitively.
